File: MoralisDotNet.BlazorUtils/MetamaskInterop.cs
namespace MoralisDotNet.BlazorUtils
{
    using System.Threading.Tasks;
    using Microsoft.JSInterop;
    using Metamask;
    using Nethereum.JsonRpc.Client.RpcMessages;
    using Newtonsoft.Json;

    public class  MetamaskInterop : IMetamaskInterop
    {
        private readonly IJSRuntime _jsRuntime;

        public MetamaskInterop(IJSRuntime jsRuntime)
        {
            _jsRuntime = jsRuntime;
        }

        public async ValueTask<string> EnableEthereumAsync()
        {
            return await _jsRuntime.InvokeAsync<string>("MetamaskInterop.EnableEthereum");
        }

        public async ValueTask<bool> CheckMetamaskAvailability()
        {
            return await _jsRuntime.InvokeAsync<bool>("MetamaskInterop.IsMetamaskAvailable");
        }

        public async ValueTask<RpcResponseMessage> SendAsync(RpcRequestMessage rpcRequestMessage)
        {
            var response = await _jsRuntime.InvokeAsync<string>("MetamaskInterop.Request", JsonConvert.SerializeObject(rpcRequestMessage));
            return JsonConvert.DeserializeObject<RpcResponseMessage>(response);
        }

        public async ValueTask<RpcResponseMessage> SendTransactionAsync(MetamaskRpcRequestMessage rpcRequestMessage)
        {
            var response = await _jsRuntime.InvokeAsync<string>("MetamaskInterop.Request", JsonConvert.SerializeObject(rpcRequestMessage));
            return JsonConvert.DeserializeObject<RpcResponseMessage>(response);
        }

        public async ValueTask<string> SignAsync(string utf8Hex)
        {
            var result = await _jsRuntime.InvokeAsync<string>("MetamaskInterop.Sign", utf8Hex);
            return result.Trim('"');
        }

        public async ValueTask<string> GetSelectedAddress()
        {
            return await _jsRuntime.InvokeAsync<string>("MetamaskInterop.GetSelectedAddress");
        }


        [JSInvokable()]
        public static async Task MetamaskAvailableChanged(bool available)
        {
            await MetamaskHostProvider.Current.ChangeMetamaskAvailableAsync(available);
        }

        [JSInvokable()]
        public static async Task SelectedAccountChanged(string selectedAccount)
        {
            await MetamaskHostProvider.Current.ChangeSelectedAccountAsync(selectedAccount);
        }

        [JSInvokable()]
        public static async Task NetworkChanged(int networkId)
        {
            await MetamaskHostProvider.Current.ChangeSelectedNetworkAsync(networkId);
        }

        [JSInvokable()]
        public static async Task OnConnect(string chainId)
        {
            await MetamaskHostProvider.Current.OnConnectAsync(chainId);
        }

        [JSInvokable()]
        public static async Task OnDisconnect(string message)
        {
            await MetamaskHostProvider.Current.OnDisconnectAsync(message);
        }

        [JSInvokable()]
        public static async Task ChainChanged(string chainId)
        {
            await MetamaskHostProvider.Current.ChainChangedAsync(chainId);
        }

        [JSInvokable()]
        public static async Task OnMessageReceived(ProviderMessage message)
        {
            await MetamaskHostProvider.Current.OnMessageReceived(message);
        }
    }
}

File: MoralisDotNet.BlazorUtils/wwwroot/MetamaskInterop.js
window.MetamaskInterop = {
    EnableEthereum: async () => {
        try {
            const accounts = await ethereum.request({ method: 'eth_requestAccounts' });
            ethereum.autoRefreshOnNetworkChange = false;
            ethereum.on("accountsChanged",
                function (accounts) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'SelectedAccountChanged', accounts[0]);
                });
            ethereum.on("networkChanged",
                function (networkId) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'NetworkChanged', networkId);
                });
            ethereum.on("connect",
                function (result) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'OnConnect', result.chainId);
                });
            ethereum.on("disconnect",
                function (result) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'OnDisconnect', result.message);
                });
            ethereum.on("chainChanged",
                function (chainid) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'ChainChanged', chainid);
                });
            ethereum.on("message",
                function (message) {
                    DotNet.invokeMethodAsync('MoralisDotNet.BlazorUtils', 'OnMessageReceived', message);
                });
            return accounts[0];
        } catch (error) {
            return null;
        }
    },
    IsMetamaskAvailable: () => {
        if (window.ethereum) return true;
        return false;
    },
    GetSelectedAddress: () => {
        return ethereum.selectedAddress;
    },

    Request: async (message) => {
        try {

            let parsedMessage = JSON.parse(message);
          
            const response = await ethereum.request(parsedMessage);
            let rpcResonse = {
                jsonrpc: "2.0",
                result: response,
                id: parsedMessage.id,
                error: null
            }
  

            return JSON.stringify(rpcResonse);
        } catch (e) {
            let rpcResonseError = {
                jsonrpc: "2.0",
                id: parsedMessage.id,
                error: {
                    message: e,
                }
            }
            return JSON.stringify(rpcResonseError);
        }
    },

    Send: async (message) => {
        return new Promise(function (resolve, reject) {
          
            ethereum.send(JSON.parse(message), function (error, result) {
                console.log(result);
                console.log(error);
                resolve(JSON.stringify(result));
            });
        });
    },

    Sign: async (utf8HexMsg) => {
        return new Promise(function (resolve, reject) {
            const from = ethereum.selectedAddress;
            const params = [utf8HexMsg, from];
            const method = 'personal_sign';
            ethereum.send({
                method,
                params,
                from,
            }, function (error, result) {
                if (error) {
                    reject(error);
                } else {
                    console.log(result.result);
                    resolve(JSON.stringify(result.result));
                }

            });
        });
    }
}